File: src/api/tauri/health.ts
/**
 * Typed Tauri IPC wrappers for Health Intelligence commands.
 *
 * These are the ONLY entry points for Tauri `invoke` calls related to health.
 * Components and pages MUST NOT call `invoke` directly (doc 48 AC-FS-04).
 *
 * Command names are snake_case (Tauri convention).
 */

import { invoke } from '@tauri-apps/api/core';
import type { HealthAlert, HealthSample } from '../../types/device.types';

/** Capture a new health sample for the local device and persist it. */
export const collectHealthSample = (): Promise<HealthSample> =>
  invoke<HealthSample>('collect_health_sample');

/** Retrieve all health samples, newest first. */
export const getHealthSamples = (): Promise<HealthSample[]> =>
  invoke<HealthSample[]>('get_health_samples');

/** Retrieve the most recent health sample, or null if none exist. */
export const getLatestHealthSample = (): Promise<HealthSample | null> =>
  invoke<HealthSample | null>('get_latest_health_sample');

/** Retrieve all health alerts, unacknowledged first then newest. */
export const getHealthAlerts = (): Promise<HealthAlert[]> =>
  invoke<HealthAlert[]>('get_health_alerts');

/** Mark a health alert acknowledged. */
export const acknowledgeAlert = (alertId: string): Promise<void> =>
  invoke<void>('acknowledge_alert', { alertId });

File: src-tauri/src/commands/health.rs
//! Health Intelligence IPC command handlers.
//!
//! Each handler locks the shared SQLite connection from [`AppState`] and
//! delegates to the `health` and `storage` layers. They contain no business
//! logic and no direct SQL (see doc 48 §4.1).

use tauri::State;

use crate::error::CoreError;
use crate::health;
use crate::models::{HealthAlert, HealthSample};
use crate::storage::{alerts_repo, health_repo};
use crate::AppState;

/// Captures a new health sample for the local device, persists it, and returns
/// it. OS sampling runs before a short-lived DB write (no long lock across I/O).
#[tauri::command]
pub fn collect_health_sample(state: State<'_, AppState>) -> Result<HealthSample, CoreError> {
    // Sample CPU/memory/disk outside any DB connection.
    let metrics = health::sampler::sample();
    let conn = state.conn()?;
    health::capture_sample_from_metrics(&conn, metrics)
}

/// Returns all health samples, newest first.
#[tauri::command]
pub fn get_health_samples(state: State<'_, AppState>) -> Result<Vec<HealthSample>, CoreError> {
    let conn = state.conn()?;
    health_repo::list_samples(&conn)
}

/// Returns the most recent health sample, or `None` if none have been recorded.
#[tauri::command]
pub fn get_latest_health_sample(
    state: State<'_, AppState>,
) -> Result<Option<HealthSample>, CoreError> {
    let conn = state.conn()?;
    health_repo::latest_sample(&conn)
}

/// Returns all health alerts, unacknowledged first then newest.
#[tauri::command]
pub fn get_health_alerts(state: State<'_, AppState>) -> Result<Vec<HealthAlert>, CoreError> {
    let conn = state.conn()?;
    alerts_repo::list_alerts(&conn)
}

/// Marks a health alert acknowledged.
#[tauri::command]
pub fn acknowledge_alert(state: State<'_, AppState>, alert_id: String) -> Result<(), CoreError> {
    let conn = state.conn()?;
    alerts_repo::acknowledge(&conn, &alert_id)?;
    Ok(())
}
